File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
n = int(input())
graph = [list(map(int, input().split())) for _ in range(n)]
visited = [False]*n
answer = 10000000


def dfs(depth, index):
    global answer
    if depth == n//2:
        pow1, pow2 = 0, 0
        for i in range(n):
            for j in range(n):
                if visited[i] and visited[j]:
                    pow1 += graph[i][j]
                elif not visited[i] and not visited[j]:
                    pow2 += graph[i][j]

        answer = min(answer, abs(pow1-pow2))
        return

    for i in range(index, n):
        if not visited[i]:
            visited[i] = True
            dfs(depth+1, i+1)
            visited[i] = False


dfs(0, 0)
print(answer)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <queue>
#include <stdio.h>
#include <string>
#include <vector>
using namespace std;
// 14:49start
int n, answer{100000000};
int graph[21][21];
bool is_red[21];

void dfs(int index, int count) {
  if (count == n / 2) {
    int total = 0, res = 0;

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (is_red[i] && is_red[j]) {
          total += graph[i][j];
        } else if (!is_red[i] && !is_red[j]) {
          res += graph[i][j];
        }
      }
    }
    answer = min(answer, abs(total - res));
    return;
  }
  for (int i = index; i < n; i++) {
    if (!is_red[i]) {
      is_red[i] = true;
      dfs(i + 1, count + 1);
      is_red[i] = false;
    }
  }
}

int main() {
  scanf("%d", &n);
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++)
      scanf("%d", &graph[i][j]);
  }
  dfs(0, 0);
  printf("%d", answer);
  return 0;
}